project-manager/src/main/scala/org/enso/projectmanager/services/TutorialsDownloader.scala

```
@@ -0,0 +1,142 @@
+)(implicit val system: ActorSystem,
+  implicit val executor: ExecutionContext,
+  implicit val materializer: ActorMaterializer)
+    extends GithubJsonProtocol {
+
+  val packagesGithubUrl =
+    s"https://api.github.com/orgs/$packagesGithubOrganisation"
+  val packagesGithubReposUrl = s"$packagesGithubUrl/repos"
+
+  implicit val logSource: LogSource[TutorialsDownloader] = _ =>
+    "TutorialsDownloader"
+  val logging = Logging(system, this)
+
+  tutorialsDir.mkdirs()
+  cacheDir.mkdirs()
+
+  def downloadUrlFor(tutorial: GithubTutorial): String =
+    s"https://github.com/$packagesGithubOrganisation/${tutorial.name}/archive/master.zip"
+
+  def zipFileFor(tutorial: GithubTutorial): File =
+    new File(cacheDir, s"${tutorial.name}.zip")
+
+  def getAvailable: Future[List[GithubTutorial]] =
+    Http()
+      .singleRequest(HttpRequest(uri = packagesGithubReposUrl))
+      .flatMap(resp => Unmarshal(resp.entity).to[String])
+      .map(JsonParser(_).convertTo[List[GithubTutorial]])
+
+  def downloadIfNeeded(tutorial: GithubTutorial): Future[File] =
+    if (needsUpdate(tutorial)) {
+      logging.info(s"Downloading zip archive for ${tutorial.name}")
+      downloadZip(tutorial)
+    }
+    else {
+      logging.info(s"Reusing cached zip archive for ${tutorial.name}")
+      Future { zipFileFor(tutorial) }
+    }
+
+  def needsUpdate(tutorial: GithubTutorial): Boolean =
+    zipFileFor(tutorial).lastModified < tutorial.lastPush
+
+  def downloadZip(tutorial: GithubTutorial): Future[File] = {
+    val downloadUrl = downloadUrlFor(tutorial)
+    val request     = HttpRequest(uri = downloadUrl)
+    val response    = HttpHelper().performWithRedirects(request)
+    val target      = zipFileFor(tutorial)
+    response
+      .flatMap(_.entity.dataBytes.runWith(FileIO.toPath(target.toPath)))
+      .map(_ => target)
+  }
+
+  def unzip(source: File): Unit = {
+    logging.debug(s"Unzipping $source\n")
+    val zipFileTry = Try(new ZipFile(source))
```

If I call `unzip` function with filepath that does not designate a zip file, I'd expect it to throw an exception rather than silently fial.